lib/cloud_controller/membership.rb

```diff
@@ -16,77 +16,112 @@ def initialize(user)
+      else
+        Organization.where(id: member_org_ids(roles))
+      end
     end
 
-    def org_guids_for_roles_subquery(roles)
-      orgs_for_roles_subquery(roles).select(:guid)
+    def authorized_space_guids(roles)
+      authorized_space_guids_subquery(roles).select_map(:guid)
     end
 
-    def orgs_for_roles_subquery(roles)
-      org_ids = org_ids_for_org_roles(roles)
-      space_ids = space_ids_for_space_roles(roles)
-      org_ids_from_space_roles = if space_ids
-                                   Space.where(id: space_ids).select(:organization_id)
-                                 end
-      Organization.where(id: org_ids).or(id: org_ids_from_space_roles).select(:id, :guid)
+    def authorized_space_guids_subquery(roles)
+      authorized_spaces_subquery(roles).select(:guid)
     end
 
-    def space_guids_for_roles(roles)
-      space_guids_for_roles_subquery(roles).select_map(:guid)
+    def authorized_spaces_subquery(roles)
+      org_ids = member_org_ids(roles)
+      if org_ids
+        Space.where(id: member_space_ids(roles)).or(organization_id: org_ids).select(:id, :guid)
+      else
+        Space.where(id: member_space_ids(roles)).select(:id, :guid)
+      end
     end
 
-    def space_guids_for_roles_subquery(roles)
-      spaces_for_roles_subquery(roles).select(:guid)
+    def member_space_ids(roles)
+      space_role_subqueries(roles).reduce do |query, subquery|
+        query.union(subquery, from_self: false)
+      end&.select_map(:space_id)
     end
 
-    def spaces_for_roles_subquery(roles)
-      space_ids = space_ids_for_space_roles(roles)
-      org_ids = org_ids_for_org_roles(roles)
-      Space.where(id: space_ids).or(organization_id: org_ids).select(:id, :guid)
+    def member_org_ids(roles)
+      org_role_subqueries(roles).reduce do |query, subquery|
+        query.union(subquery, from_self: false)
+      end&.select_map(:organization_id)
     end
 
     private
 
-    def space_roles(roles)
-      Array(roles) & SPACE_ROLES
+    def role_applies_to_space?(roles, space_id)
+      return false unless space_id && space_role_sufficient?(roles)
+
+      member_space_ids(roles).include?(space_id)
     end
 
-    def org_roles(roles)
-      Array(roles) & ORG_ROLES
+    def role_applies_to_org?(roles, org_id)
+      return false unless org_id && org_role_sufficient?(roles)
+
+      member_org_ids(roles).include?(org_id)
     end
 
-    def space_role?(roles)
-      space_roles(roles).any?
+    def roles_filter(roles, filter)
+      Array(roles).intersection(filter)
     end
 
-    def org_role?(roles)
-      org_roles(roles).any?
+    def space_role_sufficient?(roles)
+      roles_filter(roles, SPACE_ROLES).any?
     end
 
-    def space_ids_for_space_roles(roles)
-      if space_role?(roles)
-        SpaceRole.where(type: space_roles(roles), user_id: @user.id).select(:space_id)
-      end
+    def org_role_sufficient?(roles)
```

Comment:
should these methods be called something like `contains_space_role?` and `contains_org_role?`

Reply:
No problem - have adopted those names